### PR TITLE
Heretic rust rework

### PR DESCRIPTION
**Pull request by GoblinBackwards**
## About The Pull Request
Thanks to Melbert for actually coming up with the idea for this

Makes it so heretic starts off being able to rust a limited group of turfs at the start of the game, but grow to be able to rust more and more different materials as they get stronger.

Removes the 50% chance to do nothing from trying to rust reinforced walls, instead makes it so trying to rust an already rusty reinforced wall turns it into a regular rusty wall, making it take 3 hits total to break through.

Makes the particles that appear when rusting basic floors and walls now apply when rusting any turf.

Basic iron tiles no longer get replaced by plating when rusted (This just seemed like old code that got left behind from before rust became an element that could be applied to anything)

Rust walkers work at the same level as an ascended rust heretic, since they're basically beings of rust, similar to how ashmen get an ascended ash heretic power.

When playing as a non-rust path, your rust strength increases when you learn aggressive spread, entropic plume, and when you ascend, meaning you have to fulfill all three of these as a non-rust heretic to be able to rust titanium.

**Progression:**
| Rust Strength | Unlocked At | Materials that can be rusted |
| ----------- | ----------- | ----------- |
| 1 | Grasp of Rust | Basic iron walls/floors |
| 2 | Mark of Rust | Reinforced walls/floors, Plasteel |
| 3 | Toxic Blade | Titanium and Plastitanium |
| 4 | Rust Ascension | Almost anything, this is the default value |
| 5 | Never | This is for turfs that should be completely immune to rust |

## Why It's Good For The Game
Adds a new layer of progression to heretic. Removes RNG pain with trying to rust reinforced walls while still making them more of a challenge to get past. Allows rust heretic to actually function on shuttles while still having materials that you could use to trap a non-ascended heretic in a prison of silver with.
image

## Changelog

:cl:
expansion: Heretic now has its ability to rust turfs scale with knowledge.
balance: Heretics now start off with limited rusting abilities, but grow to be able to rust almost anything as they progress
balance: Reinforced walls/floors can be rusted after unlocking Mark of Rust
balance: Titanium and plastitanium can be rusted after unlocking Toxic Blade
balance: Rust heretics can spread rust to almost anything after ascending.
balance: When playing as a non-rust heretic your rusting ability improves when you learn aggressive spread, entropic plume, and when you ascend. You need at least two to rust reinforced walls, and all three to be able to rust titanium.
balance: Rusting reinforced walls is no longer RNG, and instead turns them into a regular rusted wall after being rusted twice.
/:cl:

### CODE REVIEW

File: code/game/turfs/turf.dm [hunk 1]
```diff
@@ -73,6 +73,14 @@ GLOBAL_LIST_EMPTY(station_turfs)
 	/// See __DEFINES/construction.dm for RCD_MEMORY_*.
 	var/rcd_memory
 
+	/// Minimum level of rust_strength needed to rust this tile
+	/// 1: Basic iron
+	/// 2: Reinforced floors/walls, unlocked early
+	/// 3: Titanium/Plastitanium, unlocked late
+	/// 4: Most other materials, unlocked at ascention
+	/// 5: Completely immune to being rusted
+	var/rust_resistance = 4
```
Comment (MrMelbert) [inline code/game/turfs/turf.dm:82]: These should probably be defines.
These are just ideas, if you have better ones feel free to use them:
`RUST_RESISTANCE_BASIC`
`RUST_RESISTANCE_REINFORCED`
`RUST_RESISTANCE_TITANIUM`
`RUST_RESISTANCE_ORGANIC`
`RUST_RESISTANCE_ABSOLUTE`

File: code/game/turfs/turf.dm [hunk 2]
```diff
@@ -579,9 +587,28 @@ GLOBAL_LIST_EMPTY(station_turfs)
...
+
+	if(heretic_rust_strength >= rust_resistance)
+		rust_turf()
+
+/// Override this to change behaviour when being rusted by a heretic
+/turf/proc/rust_turf()
 	if(HAS_TRAIT(src, TRAIT_RUSTY))
 		return
+	if(prob(70))
+		new /obj/effect/temp_visual/glowing_rune(src)
```
Comment (MrMelbert) [inline code/game/turfs/turf.dm:611]: Put this effect in `rust_heretic_act` after `rust_turf` is called

File: code/game/turfs/turf.dm [hunk 3]
```diff
@@ -579,9 +587,28 @@ GLOBAL_LIST_EMPTY(station_turfs)
 /turf/proc/acid_melt()
 	return
 
-/turf/rust_heretic_act()
+/// Check if the heretic is strong enough to rust this turf
+/turf/rust_heretic_act(mob/living/source)
+	var/heretic_rust_strength
```
Comment (MrMelbert) [inline code/game/turfs/turf.dm:592]: Have this set to a default value, like 4 or something.

File: code/game/turfs/turf.dm [hunk 4]
```diff
@@ -579,9 +587,28 @@ GLOBAL_LIST_EMPTY(station_turfs)
...
+	if(IS_HERETIC(source))
+		var/datum/antagonist/heretic/heretic_data = source.mind.has_antag_datum(/datum/antagonist/heretic)
+		heretic_rust_strength = heretic_data.rust_strength
+	else if(IS_HERETIC_MONSTER(source))
+		var/datum/antagonist/heretic_monster/heretic_monster_data = source.mind.has_antag_datum(/datum/antagonist/heretic_monster)
+		var/datum/antagonist/heretic/master_heretic_data = heretic_monster_data.master.has_antag_datum(/datum/antagonist/heretic)
+		heretic_rust_strength = master_heretic_data.rust_strength
+	else if(source == null) // This should only apply to rust being spread by an ascended rust heretic's rune
+		heretic_rust_strength = 4
```
Comment (MrMelbert) [inline code/game/turfs/turf.dm:601]: Also, this if statement needs to be re-arranged a bit.

Check if source is null first (also, use the helper)
```dm
if(isnull(source))
   ...
```

`IS_HERETIC` actually returns the heretic antag datum if found - you can use this to save yourself some `has_antag_datum` calls.
```dm
var/datum/antagonist/heretic/heretic_data = IS_HERETIC(source)
if(heretic_data)
   ...
```

File: code/modules/antagonists/heretic/heretic_antag.dm [hunk 5]
```diff
@@ -559,6 +561,10 @@
 /datum/antagonist/heretic/proc/get_knowledge(wanted)
 	return researched_knowledge[wanted]
 
+/datum/antagonist/heretic/proc/increase_rust_strength(strength)
+	if(rust_strength < strength)
+		rust_strength = strength
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/heretic_antag.dm:566]: - `increase_rust_strength` is a weird name for this, maybe go with like `set_rust_strength`. 
- Either check that `strength` is a number, or have strength be a default value, for sanity
```dm
if(!isnum(strength))
   return // Or crash
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 6]
```diff
@@ -66,6 +67,8 @@
 /datum/heretic_knowledge/rust_fist/on_gain(mob/user)
 	RegisterSignal(user, COMSIG_HERETIC_MANSUS_GRASP_ATTACK, .proc/on_mansus_grasp)
 	RegisterSignal(user, COMSIG_HERETIC_MANSUS_GRASP_ATTACK_SECONDARY, .proc/on_secondary_mansus_grasp)
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(1)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:71]: `our_heretic` should never ever be `null`, so you don't need the `?.` here
```suggestion
	our_heretic.increase_rust_strength(1)
```
(even if it was null for some reason, we'd want it to error because something broke)

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 7]
```diff
@@ -156,6 +160,8 @@
 /datum/heretic_knowledge/rust_mark/on_gain(mob/user)
 	RegisterSignal(user, COMSIG_HERETIC_MANSUS_GRASP_ATTACK, .proc/on_mansus_grasp)
 	RegisterSignal(user, COMSIG_HERETIC_BLADE_ATTACK, .proc/on_eldritch_blade)
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(2)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:164]: Same here
```suggestion
	our_heretic.increase_rust_strength(2)
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 8]
```diff
@@ -214,6 +227,8 @@
 
 /datum/heretic_knowledge/rust_blade_upgrade/on_gain(mob/user)
 	RegisterSignal(user, COMSIG_HERETIC_BLADE_ATTACK, .proc/on_eldritch_blade)
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(3)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:231]: Same here
```suggestion
	our_heretic.increase_rust_strength(3)
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 9]
```diff
@@ -198,9 +205,15 @@
 	cost = 1
 	route = PATH_RUST
 
+/datum/heretic_knowledge/spell/area_conversion/on_gain(mob/user)
+	. = ..()
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(2)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:211]: Same here
```suggestion
	our_heretic.increase_rust_strength(2)
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 10]
```diff
@@ -240,13 +256,19 @@
 	cost = 1
 	route = PATH_RUST
 
+/datum/heretic_knowledge/spell/entropic_plume/on_gain(mob/user)
+	. = ..()
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(3)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:262]: Same here
```suggestion
	our_heretic.increase_rust_strength(3)
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 11]
```diff
@@ -294,6 +316,8 @@
 	RegisterSignal(user, COMSIG_MOVABLE_MOVED, .proc/on_move)
 	RegisterSignal(user, COMSIG_LIVING_LIFE, .proc/on_life)
 	user.client?.give_award(/datum/award/achievement/misc/rust_ascension, user)
+	var/datum/antagonist/heretic/our_heretic = IS_HERETIC(user)
+	our_heretic?.increase_rust_strength(4)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:320]: Same here
```suggestion
	our_heretic.increase_rust_strength(4)
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 12]
```diff
@@ -363,7 +387,7 @@
 
 /datum/rust_spread/New(loc)
 	centre = get_turf(loc)
-	centre.rust_heretic_act()
+	centre.rust_heretic_act(null)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:390]: `null` is implicit by not passing any parameters, not needed
```suggestion
	centre.rust_heretic_act()
```

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 13]
```diff
@@ -384,7 +408,7 @@
 		if(!length(edge_turfs))
 			break
 		var/turf/afflicted_turf = pick_n_take(edge_turfs)
-		afflicted_turf.rust_heretic_act()
+		afflicted_turf.rust_heretic_act(null)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:411]: Ditto
```suggestion
		afflicted_turf.rust_heretic_act()
```

File: code/modules/antagonists/heretic/magic/rust_wave.dm [hunk 14]
```diff
@@ -19,7 +19,7 @@
 
 /obj/effect/proc_holder/spell/cone/staggered/entropic_plume/do_turf_cone_effect(turf/target_turf, level)
 	. = ..()
-	target_turf.rust_heretic_act()
+	target_turf.rust_heretic_act(usr)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/magic/rust_wave.dm:22]: Have you tested this / are you sure that `usr` is set correctly for this proc? You may need to adjust cone spells a bit if not.

**Review by MrMelbert — CHANGES_REQUESTED** [review]
There's some things i'm not entirely sold on with this yet. 
Like i'm not sure if non-rust heretics should be allowed to cross-path into entropic plume / aggressive spread to jump ahead and be able to rust as good as rust heretics themselves. 

Otherwise this is a preliminary code review

File: code/modules/antagonists/heretic/magic/rust_wave.dm (hunk 14 above)
Comment (GoblinBackwards) [reply code/modules/antagonists/heretic/magic/rust_wave.dm:22]: Yeah this works fine.

**Comment by GoblinBackwards** [comment]
> Like i'm not sure if non-rust heretics should be allowed to cross-path into entropic plume / aggressive spread to jump ahead and be able to rust as good as rust heretics themselves.

Made some changes with how they gain rust strength when playing non-rust paths that should help with it. I made rust strength increment when learning entropic plume/aggressive spread, and when ascending. So for a non-rust path to be able to reach the point of rusting titanium they'd need to research both rust spells as well as ascend, instead of before where they could skip there just by buying entropic plume.

File: code/modules/antagonists/heretic/magic/aggressive_spread.dm [hunk 15]
```diff
@@ -19,7 +19,7 @@
 		var/chance = 100 - (max(get_dist(T,user),1)-1)*100/(range+1)
 		if(!prob(chance))
 			continue
-		T.rust_heretic_act()
+		T.rust_heretic_act(usr)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/magic/aggressive_spread.dm:22]:
```suggestion
		T.rust_heretic_act(user)
```
Man I hate proc holders

File: code/game/turfs/turf.dm [hunk 16]
```diff
@@ -579,9 +581,24 @@ GLOBAL_LIST_EMPTY(station_turfs)
 /turf/proc/acid_melt()
 	return
 
-/turf/rust_heretic_act()
+/// Check if the heretic is strong enough to rust this turf
+/turf/rust_heretic_act(mob/living/source)
 	if(turf_flags & NO_RUST)
 		return
```
Comment (MrMelbert) [inline code/game/turfs/turf.dm:587]: Put this in `rust_turf()` since the other PR got merged anyways

File: code/modules/antagonists/heretic/knowledge/ash_lore.dm [hunk 17]
```diff
@@ -223,7 +223,8 @@
 		When completed, you become a harbinger of flames, gaining two abilites. \
 		Cascade, which causes a massive, growing ring of fire around you, \
 		and Oath of Flame, causing you to passively create a ring of flames as you walk. \
-		You will also become immune to flames, space, and similar environmental hazards."
+		You will also become immune to flames, space, and similar environmental hazards. \
+		Also improves your ability to spread rust."
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/ash_lore.dm:227]: I don't like putting "Also improves your ability to spread rust." in all of the other ascensions - too wordy and no one really cares if you're not already a rust heretic

File: code/modules/antagonists/heretic/knowledge/rust_lore.dm [hunk 18]
```diff
@@ -186,7 +192,8 @@
 /datum/heretic_knowledge/spell/area_conversion
 	name = "Aggressive Spread"
 	desc = "Grants you Aggressive Spread, a spell that spreads rust to nearby surfaces. \
-		Already rusted surfaces are destroyed."
+		Already rusted surfaces are destroyed. \
+		Improves your rusting abilities if learned as a side path."
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/knowledge/rust_lore.dm:196]: Maybe instead "Improves your rusting abilities for non-rust heretics"?

File: code/modules/antagonists/heretic/heretic_antag.dm [hunk 19]
```diff
@@ -559,6 +561,12 @@
 /datum/antagonist/heretic/proc/get_knowledge(wanted)
 	return researched_knowledge[wanted]
 
+/datum/antagonist/heretic/proc/increase_rust_strength(side_path_only = FALSE)
```
Comment (MrMelbert) [inline code/modules/antagonists/heretic/heretic_antag.dm:564]: Consider making a setter for this as well - so you can directly set strength to a certain level (r-wall level or something)
